lsp/nls/src/position.rs

```diff
@@ -0,0 +1,162 @@
+use codespan::ByteIndex;
+use nickel_lang_core::{
+    position::TermPos,
+    term::{RichTerm, SharedTerm, Traverse, TraverseControl},
+};
+use range_map::{Range, RangeMap};
+
+// A term that uses pointer equality to implement Eq.
+#[derive(Clone, Debug)]
+struct RichTermPtr(RichTerm);
+
+impl PartialEq for RichTermPtr {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.pos == other.0.pos && SharedTerm::ptr_eq(&self.0.term, &other.0.term)
+    }
+}
+
+impl Eq for RichTermPtr {}
+
+/// A lookup data structure, for looking up the term at a given position.
+///
+/// Overlapping positions are resolved in favor of the smaller one; i.e., lookups return the
+/// most specific term for a given position.
+#[derive(Clone, Debug)]
+pub struct PositionLookup {
+    table: RangeMap<u32, RichTermPtr>,
+}
+
+impl PositionLookup {
+    /// Create a position lookup table for looking up subterms of `rt` based on their positions.
+    pub fn new(rt: &RichTerm) -> Self {
+        let mut all_ranges = Vec::new();
+
+        rt.traverse_ref(&mut |term: &RichTerm| {
```

Comment:
It definitely could (it would have to have an extra `fn skip_subtree(&mut self)` method). I went with the recursive version because the state tracking was easier (i.e. non-existent), but I could take a stab at the iterator version also.